x/avs/keeper/task.go

```diff
@@ -102,3 +110,264 @@ func (k Keeper) IterateTaskAVSInfo(ctx sdk.Context, fn func(index int64, taskInf
+		//if err != nil || info.TaskId != resp.TaskID {
+		//	return errorsmod.Wrap(
+		//		types.ErrParamError,
+		//		fmt.Sprintf("SetTaskResultInfo: invalid param value:%s", info.Stage),
+		//	)
+		//}
+		// check bls sig
+		flag, err := blst.VerifySignature(info.BlsSignature, taskResponseDigest, pubKey)
+		if !flag || err != nil {
+			return errorsmod.Wrap(
+				types.ErrSigVerifyError,
+				fmt.Sprintf("SetTaskResultInfo: invalid task address: %s (Task ID: %d)", info.TaskContractAddress, info.TaskId),
+			)
+		}
+
+		infoKey := assetstype.GetJoinedStoreKey(info.OperatorAddress, info.TaskContractAddress, strconv.FormatUint(info.TaskId, 10))
+
+		store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixTaskResult)
+		bz := k.cdc.MustMarshal(info)
+		store.Set(infoKey, bz)
+		return nil
+	default:
+		return errorsmod.Wrap(
+			types.ErrParamError,
+			fmt.Sprintf("SetTaskResultInfo: invalid param value:%s", info.Stage),
+		)
+	}
+}
+
+func (k *Keeper) IsExistTaskResultInfo(ctx sdk.Context, operatorAddress, taskContractAddress string, taskID uint64) bool {
+	infoKey := assetstype.GetJoinedStoreKey(operatorAddress, taskContractAddress,
+		strconv.FormatUint(taskID, 10))
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixTaskResult)
+	return store.Has(infoKey)
+}
+
+func (k *Keeper) GetTaskResultInfo(ctx sdk.Context, operatorAddress, taskContractAddress string, taskID uint64) (info *types.TaskResultInfo, err error) {
+	if !common.IsHexAddress(taskContractAddress) {
+		return nil, types.ErrInvalidAddr
+	}
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixTaskResult)
+	infoKey := assetstype.GetJoinedStoreKey(operatorAddress, taskContractAddress,
+		strconv.FormatUint(taskID, 10))
+	value := store.Get(infoKey)
+	if value == nil {
+		return nil, errorsmod.Wrap(types.ErrNoKeyInTheStore,
+			fmt.Sprintf("GetTaskResultInfo: key is %s", infoKey))
+	}
+
+	ret := types.TaskResultInfo{}
+	if err := k.cdc.Unmarshal(value, &ret); err != nil {
+		return nil, errorsmod.Wrap(err, "GetTaskResultInfo: failed to unmarshal task result info")
+	}
+	return &ret, nil
+}
+
+// IterateResultInfo iterate through task result info
+func (k Keeper) IterateResultInfo(ctx sdk.Context, fn func(index int64, info types.TaskResultInfo) (stop bool)) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixTaskResult)
+
+	iterator := sdk.KVStorePrefixIterator(store, nil)
+	defer iterator.Close()
+
+	i := int64(0)
+
+	for ; iterator.Valid(); iterator.Next() {
+		task := types.TaskResultInfo{}
+		k.cdc.MustUnmarshal(iterator.Value(), &task)
+
+		stop := fn(i, task)
+
+		if stop {
+			break
+		}
+		i++
+	}
+}
+
+func (k Keeper) GroupTasksByIDAndAddress(tasks []types.TaskResultInfo) map[string][]types.TaskResultInfo {
+	taskMap := make(map[string][]types.TaskResultInfo)
+	for _, task := range tasks {
+		key := task.TaskContractAddress + "_" + strconv.FormatUint(task.TaskId, 10)
+		taskMap[key] = append(taskMap[key], task)
```

Comment:
since the order of iterating over a map is not deterministic, would this lead to non-deterministic list of TaskResultInfo? since the order may vary for each iteration

Reply:
This method is only used for grouping and does not require any order

Reply:
but later in `AfterEpochEnd` hook, this list would be iterated over to get signedOperatorList and diff, and assign them to taskinfo's fields. Given taskinfo would be persisted, if signedOperatorList or diff does not have determined order, it could result in apphash mismatch?